localstack/runtime/shutdown.py

```diff
@@ -0,0 +1,49 @@
+from typing import Any, Callable
+
+from localstack.runtime import hooks
+from localstack.utils.functions import call_safe
+
+
+class ShutdownHandlers:
+    """
+    Register / unregister shutdown handlers. All registered shutdown handlers should execute as fast as possible.
+    Blocking shutdown handlers will block infra shutdown.
+    """
+
+    def __init__(self):
+        self._callbacks = []
+
+    def register(self, shutdown_handler: Callable[[], Any]) -> None:
+        """
+        Register shutdown handler. Handler should not block or take more than a couple seconds.
+
+        :param shutdown_handler: Callable without parameters
+        """
+        self._callbacks.append(shutdown_handler)
+
+    def unregister(self, shutdown_handler: Callable[[], Any]) -> None:
+        """
+        Unregister a handler. Idempotent operation.
+
+        :param shutdown_handler: Shutdown handler which was previously registered
+        """
+        try:
+            self._callbacks.remove(shutdown_handler)
+        except ValueError:
+            pass
+
+    def run(self) -> None:
+        """
+        Execute shutdown handlers in reverse order of registration.
+        Should only be called once, on shutdown.
+        """
+        for callback in reversed(list(self._callbacks)):
```

Comment:
what's the reason for wrapping it in `list` here?

Reply:
Some handlers will remove themselves from the list while they are called. To prevent some concurrent modifications while iterating, I added the list copy before, to prevent any issues should they arrive (since we are only deleting the current object, and not the next one, it should work, but better safe than sorry)